### ISSUE TITLE
Truck Factor

### ISSUE THREAD

**Issue by mircealungu** [git-truck/git-truck#587]
Dear folks,

My colleague Helge pointed me to the fact that what is now named Truck Factor view is misleading. There is an actual formula for that...

One simple solution could be that we rename the current view into 'number of developers'. And we use some kind of gradient scale as @leonelmerino  suggested for the last changed.

Although there are problems with that too... We are still not showing who is the sole developer... And that was another issue here i think...

What is important i think it's that we either discuss quickly about how to fix this, or we remove it for now so we don't confuse people who are used with the formula.

M.

**Comment by joglr** [comment]
It is true that you usually define the truck factor as a per project basis, whereas we have made our own term that is file based.

**Comment by mircealungu** [comment]
Indeed. So then maybe the simple solution would be to: 
- rename the view to "Number of contributors"
- change the color mapping to more pleasant than lots of red? I envision something like: white for a single developer and then logarithmically darker for 2, 4, 8, etc. as we do for last-changed?

**Comment by joglr** [comment]
> Indeed. So then maybe the simple solution would be to: 
> - rename the view to "Number of contributors"
> - change the color mapping to more pleasant than lots of red? I envision something like: white for a single developer and then logarithmically darker for 2, 4, 8, etc. as we do for last-changed? 

Yeah that would be cool, @emiljapelt is this something you would like to work on?

**Comment by leonelmerino** [comment]
Sure. Perhaps, the metric could not only consider the number of contributors per file but measure the median of the number of contributors for each line of code in the file. It would be interesting to distinguish files in which there are many contributors, but most lines were developed by a single one.


From: Mircea Filip Lungu ***@***.***>
Date: Sunday, 16 October 2022 at 06:34
To: git-truck/git-truck ***@***.***>
Cc: leonelmerino ***@***.***>, Mention ***@***.***>
Subject: Re: [git-truck/git-truck] Truck Factor (Issue #587)

Indeed. So then maybe the simple solution would be to:

  *   rename the view to "Number of contributors"
  *   change the color mapping to more pleasant than lots of red? I envision something like: white for a single developer and then logarithmically darker for 2, 4, 8, etc. as we do for last-changed?

**Comment by mircealungu** [comment]
> Sure. Perhaps, the metric could not only consider the number of contributors per file but measure the median of the number of contributors for each line of code in the file.

The median of the number of contributors for each line…
I’m not sure I follow. How would you compute that? 

But overall, I see the value of trying to think deeper than
just contributor count. Because of situations like the following
would have a large contributor count, but not mean much: 

99% 0.2% 0.2% 0.2% 0.2% 0.2%

So we would have to introduce something like a collaboration
index which would make the above example be lower in 
collaboration than the one below: 

33% 33% 33% 

How would we do that?

Maybe something like the gini index? Although the problem 
with the gini index is that a lot of “poor” contributors to a file 
would result in reporting inequality… 

So maybe something else, like: number of developers required
to contribute to more than 90% of the changes? 

99% 0.2% 0.2% 0.2% 0.2% 0.2% => 1

33% 33% 33% >=3 

80, 10, 2, 1, 1, 1, ... => 2 

Would this be a good metric? 


> It would be interesting to distinguish files in which there are many contributors, but most lines were developed by a single one.

Or maybe your metric that I didn’t understand @leonelmerino :)
